torch/onnx/symbolic_opset9.py

```diff
@@ -412,103 +411,114 @@ def _div_rounding_mode(g: jit_utils.GraphContext, self, other, rounding_mode):
 
-@_beartype.beartype
-def _trunc_divide(g: jit_utils.GraphContext, self, other):
-    out = g.op("Div", self, other)
-    # the correct operation is truncate, which is not supported in ONNX,
-    # we cannot call floor since it will behave differently for negative numbers
-    # (eg. -0.1 should become -0 )
-    # - if scalar_type information are not available, assume that
-    # we need to call floor (treat as float)
-    out = g.op("Cast", out, to_i=_C_onnx.TensorProtoDataType.INT64)
+def _determine_division_output_type(
+    self, other, possible_int_output=True
+) -> _type_utils.JitScalarType:
+    """Determines the output type of div.
 
-    # Matching PyTorch's behavior:
-    # - if self is fp the output's type is self's type
-    # - if self is not fp and other is fp, the output is of type JitScalarType.FLOAT
-    # - self is not fp and other is not fp, the output's type is self's output type
-    # - the output type defaults to Float
-    scalar_type = _type_utils.JitScalarType.from_value(
+    Args:
+        self: The self value.
+        other: The other value.
+        possible_int_output: Whether the output can be integer. If false, the output
+            type is at least torch.get_default_dtype().
+    """
+    # - If both self's and other's types are known, promote type based on torch's rules
+    # - If one of the types is not known, assume it is int and run type promotion
+    # - The output type defaults to default_dtype()
+    self_type = _type_utils.JitScalarType.from_value(
         self, _type_utils.JitScalarType.UNDEFINED
     )
-    if scalar_type != _type_utils.JitScalarType.UNDEFINED:
-        if not symbolic_helper._is_fp(self) and symbolic_helper._is_fp(other):
-            out = g.op("Cast", out, to_i=_C_onnx.TensorProtoDataType.FLOAT)
-        else:
-            out = g.op(
-                "Cast",
-                out,
-                to_i=scalar_type.onnx_type(),
+    other_type = _type_utils.JitScalarType.from_value(
+        other, _type_utils.JitScalarType.UNDEFINED
+    )
+
+    if (
+        self_type != _type_utils.JitScalarType.UNDEFINED
+        and other_type != _type_utils.JitScalarType.UNDEFINED
+    ):
+        out_type = torch.promote_types(self_type.dtype(), other_type.dtype())
+        if not possible_int_output and out_type not in {
+            torch.float16,
+            torch.float32,
+            torch.float64,
+            torch.bfloat16,
+        }:
+            # Convert int types to the default float type.
+            out_type = torch.promote_types(out_type, torch.get_default_dtype())
+        return _type_utils.JitScalarType.from_dtype(out_type)
+    elif (
+        self_type != _type_utils.JitScalarType.UNDEFINED
+        and other_type == _type_utils.JitScalarType.UNDEFINED
+    ):
+        return _type_utils.JitScalarType.from_dtype(
+            torch.promote_types(
+                self_type.dtype(),
+                torch.int8 if possible_int_output else torch.get_default_dtype(),
```

Comment:
Why is it `int8`?

Reply:
I used int8 as the lowest int type so that if the type is int it is retained. I updated the code to make logic clearer and got rid of int8.